scripts/kilt/anserini_retriever.py

```diff
@@ -0,0 +1,155 @@
+import multiprocessing
+from multiprocessing.pool import ThreadPool
+
+from tqdm import tqdm
+from pyserini.search import SimpleSearcher
+
+import kilt.kilt_utils as utils
+from kilt.retrievers.base_retriever import Retriever
+
+import jnius
+
+from nltk import bigrams, word_tokenize, SnowballStemmer
+from nltk.corpus import stopwords
+import string
+
+ent_start_token = "[START_ENT]"
+ent_end_token = "[END_ENT]"
+STOPWORDS = set(stopwords.words('english') + list(string.punctuation))
+stemmer = SnowballStemmer("english")
+
+
+def parse_hits(hits):
+    doc_ids = []
+    doc_scores = []
+    for hit in hits:
+        wikipedia_id = hit.docid.split('-')[0]
+        if wikipedia_id and wikipedia_id not in doc_ids:
+            doc_ids.append(wikipedia_id)
+            doc_scores.append(hit.score)
+    return doc_ids, doc_scores
+
+
+def _get_predictions_thread(arguments):
+
+    id = arguments["id"]
+    queries_data = arguments["queries_data"]
+    topk = arguments["topk"]
+    ranker = arguments["ranker"]
+    logger = arguments["logger"]
+    use_bigrams = arguments["use_bigrams"]
+    stem_bigrams = arguments["stem_bigrams"]
+
+    if id == 0:
+        iter_ = tqdm(queries_data)
+    else:
+        iter_ = queries_data
+
+    result_doc_ids = []
+    result_doc_scores = []
+    result_query_id = []
+
+    for query_element in iter_:
+
+        query = (
+            query_element["query"]
+            .replace(ent_start_token, "")
+            .replace(ent_end_token, "")
+            .strip()
+        )
+        result_query_id.append(query_element["id"])
+
+        doc_ids = []
+        doc_scores = []
+
+        if use_bigrams:
+            tokens = filter(lambda word: word not in STOPWORDS, word_tokenize(query))
+            if stem_bigrams:
+                tokens = map(stemmer.stem, tokens)
+            bigram_query = bigrams(tokens)
+            bigram_query = " ".join(["".join(bigram) for bigram in bigram_query])
+            query += " " + bigram_query
+        try:
+            hits = ranker.search(query, k=topk)
+            doc_ids, doc_scores = parse_hits(hits)
+            # doc_ids = [hit.docid for hit in hits]
+            # doc_scores = [hit.score for hit in hits]
+
+        except RuntimeError as e:
+            if logger:
+                logger.warning("RuntimeError: {}".format(e))
+        except jnius.JavaException as e:
+            if logger:
+                logger.warning("{query} jnius.JavaException: {}".format(query_element, e))
+            if 'maxClauseError' in str(e):
+                query = " ".join(query.split()[:950])
```

Comment:
A hacky way to handle maxClauseError (1024), which happens mostly for Entity Linking tasks (inputs are entire paragraphs)